src/util/importJSON.ts

```diff
@@ -0,0 +1,160 @@
+import _ from 'lodash'
+import { Child, Context, Lexeme, ParentEntry, Path } from '../types'
+import { State } from './initialState'
+import { EM_TOKEN, ROOT_TOKEN } from '../constants'
+import { getRankAfter, getThought, getThoughts, nextSibling } from '../selectors'
+
+// util
+import {
+  addThought,
+  contextOf,
+  createId,
+  equalPath,
+  equalThoughtRanked,
+  hashContext,
+  hashThought,
+  head,
+  headRank,
+  pathToContext,
+  removeContext,
+  rootedContextOf,
+  timestamp,
+} from '../util'
+import { Block } from '../action-creators/importText'
+import { GenericObject } from '../utilTypes'
+
+interface ImportHtmlOptions {
+  skipRoot? : boolean,
+}
+
+/** Recursively calculate rank offset for a given position based on thoughts before and their children. Increment offset each time we pop out of context. */
+const getRankOffset = (thoughts: Block[], position?: number): number => {
+  const before = !position ? [...thoughts] : thoughts.slice(0, position)
+  const result = before.map(thought => thought.children.length > 0 ? 1 + getRankOffset(thought.children) : 1).reduce((acc, val) => acc + val, 0)
+  return position ? result : result + 1
+}
+
+/** Prepare thoughts for saving, skip root if necessary. */
+const prepare = (skipRoot: boolean | undefined, thoughts: Block[]) => {
+  if (!skipRoot) return thoughts
+  const head = _.head(thoughts)
+  if (!head) return thoughts
+  const tail = _.tail(thoughts)
+  return head.children.length > 0 ? [...head.children, ...tail] : tail
+}
+
+/** Calculate last thought of the first level, as this is where the selection will be restored to. */
+const calculateLastThoughtFirstLevel = (rankStart: number, rankIncrement: number, thoughtsJSON: Block[]) => {
+  const lastThoughtFirstLevelIndex = thoughtsJSON.length - 1
+  const lastThoughtFirstLevel = thoughtsJSON[lastThoughtFirstLevelIndex]
+  const rankOffset = lastThoughtFirstLevelIndex === 0 ? 0 : getRankOffset(thoughtsJSON, lastThoughtFirstLevelIndex)
+  return { value: lastThoughtFirstLevel.scope, rank: rankStart + rankOffset * rankIncrement }
+}
+
+/** Recursively iterate through thoughtsJSON and call insertThought for each thought individually to save it. */
+const saveThoughts = (context: Context, rank: number, rankIncrement: number, thoughtsJSON: Block[], insertThought: (value: string, context: Context, rank: number) => void) => {
+  thoughtsJSON.forEach((thought, index) => {
+    const rankOffset = index === 0 ? 0 : getRankOffset(thoughtsJSON, index)
```

Comment:
It looks like `getRankOffset` is called for each thought in `O(n)`, and `getRankOffset` itself loops through each thought in `O(n)` recursively, resulting in computational complexity somewhere around `O(n^2*m)`.

Please confirm and suggest a solution with more acceptable performance characteristics.

Reply:
Yes, it is. I think we can calculate offsets directly in `saveThoughts` or before it, and save them to Map

Reply:
I've replaced it with one calculation for all thoughts. Check 5643990